src/crimson/os/seastore/scanner.h

```diff
@@ -0,0 +1,112 @@
+  read_segment_header_ret read_segment_header(segment_id_t segment);
+
+  /**
+   * scan_extents
+   *
+   * Scans records beginning at addr until the first record boundary after
+   * addr + bytes_to_read.
+   *
+   * Returns list<extent, extent_info>
+   * cursor.is_complete() will be true when no further extents exist in segment.
+   */
+  using scan_extents_cursor = scan_valid_records_cursor;
+  using scan_extents_ertr = read_ertr::extend<crimson::ct_error::enodata>;
+  using scan_extents_ret_bare = std::list<std::pair<paddr_t, extent_info_t>>;
+  using scan_extents_ret = scan_extents_ertr::future<scan_extents_ret_bare>;
+  scan_extents_ret scan_extents(
+    scan_extents_cursor &cursor,
+    extent_len_t bytes_to_read
+  );
+
+  using scan_valid_records_ertr = read_ertr::extend<crimson::ct_error::enodata>;
+  using scan_valid_records_ret = scan_valid_records_ertr::future<
+    size_t>;
+  using found_record_handler_t = std::function<
+    scan_valid_records_ertr::future<>(
+      paddr_t record_block_base,
+      // callee may assume header and bl will remain valid until
+      // returned future resolves
+      const record_header_t &header,
+      const bufferlist &bl)>;
+  scan_valid_records_ret scan_valid_records(
+    scan_valid_records_cursor &cursor, ///< [in, out] cursor, updated during call
+    segment_nonce_t nonce,             ///< [in] nonce for segment
+    size_t budget,                     ///< [in] max budget to use
+    found_record_handler_t &handler    ///< [in] handler for records
+  ); ///< @return used budget
+
+  /// read record metadata for record starting at start
+  using read_validate_record_metadata_ertr = read_ertr;
+  using read_validate_record_metadata_ret =
+    read_validate_record_metadata_ertr::future<
+      std::optional<std::pair<record_header_t, bufferlist>>
+    >;
+  read_validate_record_metadata_ret read_validate_record_metadata(
+    paddr_t start,
+    segment_nonce_t nonce);
+
+  /// attempts to decode extent infos from bl, return nullopt if unsuccessful
+  std::optional<std::vector<extent_info_t>> try_decode_extent_infos(
+    record_header_t header,
+    const bufferlist &bl);
+
+  /// read and validate data
+  using read_validate_data_ertr = read_ertr;
+  using read_validate_data_ret = read_validate_data_ertr::future<bool>;
+  read_validate_data_ret read_validate_data(
+    paddr_t record_base,
+    const record_header_t &header  ///< caller must ensure lifetime through
+                                   ///  future resolution
+  );
+
+  using init_segments_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  using init_segments_ret_bare =
+    std::vector<std::pair<segment_id_t, segment_header_t>>;
+  using init_segments_ret = init_segments_ertr::future<init_segments_ret_bare>;
+  init_segments_ret init_segments();
+
+  /// validate embedded metadata checksum
+  static bool validate_metadata(const bufferlist &bl);
+
+
+private:
+  SegmentManager& segment_manager;
```

Comment:
I don't think we need this either.  Scanner can simply be a namespace with a few public free functions which take a SegmentManager as a parameter -- that'll make adding more devices later easier anyway.

Reply:
Actually, I'm developping multi-device support now, and I also find that this part needs to be modified. I changed this part as follows: since paddr_t already encapsulate device_id_t, so I didn't change other methods' parameter list, only changed this field to be `std::map<device_id_t, SegmentManager*>` and made each method look for the corresponding segment manager by the paddr_t param they accepted. Since we only support 16 devices at most (top 4 bits of segment_id_t), I think the lookup shouldn't cause much cpu overhead. Is this a possible way out? Thanks:-)

Reply:
I still don't see a reason to have that mapping within Scanner.  It seems simpler to let scanner.h/cc simply contain the logic for interpreting a segment.  Its callers must already have such a mapping.

Reply:
It seems that, with EPM in place, all `paddr_t` values are provided either by the Journal or EPM, other components just carry those values around and doesn't touch it. So I think the callers doesn't have to have such a mapping, only EPM has to. Am I correct?

Reply:
I mean if we have that mapping in Scanner, only Scanner and EPM has to have such a mapping, other components won't be affected.

Reply:
If EPM is supposed to have the segment->device mapping, then I really don't want to duplicate it here (particularly since otherwise there would be no state associated with Scanner at all).

The only users of Scanner at this time are Journal and SegmentCleaner.  Journal needs a SegmentManager for the device it's using, so that leaves SegmentCleaner.

Previously, SegmentCleaner::ExtentCallbackInterface provided scan_extents as a way to abstract the Journal segment scanning machinery from SegmentCleaner.  Scanner sort of seems to be serving that purpose now in that it allows SegmentCleaner to scan a segment without maintaining a segment->SegmentManager mapping, but at the cost of requiring duplicate state between Scanner and EPM.

I think the simplest solution is to remove Scanner from SegmentCleaner and restore SegmentCleaner::ExtentCallbackInterface::scan_extents.  TransactionManager's scan_extents implementation would use EPM to get the SegmentManager reference and pass it along with the relevant Scanner function.

SegmentCleaner::init_segments (previously housed in Journal::find_replay_segments) is a bit murky because it involves segments owned by both the EPM and the Journal.  I think that EPM should provide a method to enumerate all segments (with relevant metadata -- mainly whether the segment is a Journal segment) that TransactionManager can use for Journal::replay() and to initialize SegmentCleaner.

Reply:
Um, sorry, I didn't remember it right. EPM doesn't have that mapping, it create paddr_t by getting segment id from the corresponding segment manager, which already have the device id encapsulated, and mix it with the current of offset. So, only Scanner have to hold the mapping.

Sorry for this inconvenience...

Reply:
Hmm, ok.  I'm skeptical about Scanner as implemented here being the right place for that mapping, but we can address it later if it becomes inconvenient.  I'd expect the multidevice support PR to have a single component responsible for knowing about all devices and routing reads etc.

Reply:
In that case, I'll check back once you've cleaned up init_segments and other comments.

Reply:
> Hmm, ok. I'm skeptical about Scanner as implemented here being the right place for that mapping, but we can address it later if it becomes inconvenient. I'd expect the multidevice support PR to have a single component responsible for knowing about all devices and routing reads etc.

Yes, total agree with you:-)



> In that case, ping me once you've cleaned up init_segments and the other comments.

Ok, no problem:=)